examples/tests/lr1110/lorawan/main_lorawan.c

```diff
@@ -0,0 +1,460 @@
+    uint32_t sleep_time_ms = smtc_modem_run_engine( );     // cause process fault
+
+    /* go in low power */
+    hal_mcu_set_sleep_for_ms(sleep_time_ms);
+
+  }
+}
+
+/*
+ * -----------------------------------------------------------------------------
+ * --- PRIVATE FUNCTIONS DEFINITION --------------------------------------------
+ */
+
+static void on_modem_reset(uint16_t reset_count) {
+  printf("on_modem_reset (%i)\n", reset_count);
+
+  HAL_DBG_TRACE_INFO("Application parameters:\n");
+  HAL_DBG_TRACE_INFO("  - LoRaWAN uplink Fport = %d\n", LORAWAN_APP_PORT);
+  HAL_DBG_TRACE_INFO("  - DM report interval   = %d\n", APP_TX_DUTYCYCLE);
+  HAL_DBG_TRACE_INFO("  - Confirmed uplink     = %s\n", (LORAWAN_CONFIRMED_MSG_ON == true) ? "Yes" : "No");
+
+  lorawan_apps_modem_common_configure_lorawan_params();
+
+  // We use TTN so only enable bank 2 channels.
+  region_us_915_the_things_network_init();
+
+  ASSERT_SMTC_MODEM_RC(smtc_modem_join_network(stack_id));
+}
+
+static void on_modem_network_joined(void) {
+  printf("on_modem_network_joined successful!\n");
+
+  ASSERT_SMTC_MODEM_RC(smtc_modem_alarm_start_timer(APP_TX_DUTYCYCLE));
+
+  ASSERT_SMTC_MODEM_RC(smtc_modem_adr_set_profile(stack_id, LORAWAN_DEFAULT_DATARATE, adr_custom_list));
+}
+
+static void on_modem_join_fail(void) {
+  printf("join failed!\n");
+}
+
+static void on_modem_alarm(void) {
+  printf("on_modem_alarm\n");
+  smtc_modem_status_mask_t modem_status;
+  // uint32_t                 charge        = 0;
+  uint8_t app_data_size = 0;
+
+  /* Schedule next packet transmission */
+  ASSERT_SMTC_MODEM_RC(smtc_modem_alarm_start_timer(APP_TX_DUTYCYCLE));
+  HAL_DBG_TRACE_PRINTF("smtc_modem_alarm_start_timer: %d s\n\n", APP_TX_DUTYCYCLE);
+
+  ASSERT_SMTC_MODEM_RC(smtc_modem_get_status(stack_id, &modem_status));
+  modem_status_to_string(modem_status);
+
+  // ASSERT_SMTC_MODEM_RC( smtc_modem_get_charge( &charge ) );
+
+  printf("[Sensors] Sampling Temperature and Humidity sensors once.\n");
+
+  bool temperature_available = driver_exists(DRIVER_NUM_TEMPERATURE);
+  bool humidity_available    = driver_exists(DRIVER_NUM_HUMIDITY);
+  int temp = 0;
+  int humi = 0;
+  if (temperature_available) {
+    libtocksync_temperature_read(&temp);
+    printf("Temperature: %d.%d deg C, send %d to TTN\n", temp / 100, temp % 100, temp);
+  } else {
+    printf("Temperature sensor not available.\n");
+  }
+
+  if (humidity_available) {
+    libtocksync_humidity_read(&humi);
+    printf("Humidity: %d.%d%%, send %d to TTN\n", humi / 100, humi % 100, humi);
+  } else {
+    printf("Humidity sensor not available.\n");
+  }
+
+  app_data_buffer[app_data_size++] = (uint8_t)((temp >> 8) & 0xFF);
+  app_data_buffer[app_data_size++] = (uint8_t)(temp & 0xFF);
+  app_data_buffer[app_data_size++] = (uint8_t)((humi >> 8) & 0xFF);
+  app_data_buffer[app_data_size++] = (uint8_t)(humi & 0xFF);
+
+  send_frame(app_data_buffer, app_data_size, LORAWAN_CONFIRMED_MSG_ON);
```

Comment:
This is how I started out as well, the problem is TTN can't easily decode it.

You should look at using the Cayenne Low Power Payload (LPP) encoder so then TTN will display the actual results

Reply:
I wrote local scripts to decode data from TTN, and LLP encoder sounds much more convenient